libs/base/ui/form/multi-range/src/multi-range.component.ts

```diff
@@ -13,142 +12,123 @@ export class MultiRangeComponent
     this.style.setProperty(
       '--_multi-range-max',
       `${100 - sliderMaxPercentage}%`
     );
+  }
+
+  protected syncNativeInputValues(minValue: number, maxValue: number): void {
+    this._min = minValue;
+    this._max = maxValue;
 
-    super.update(changedProperties);
+    if (this.inputMin) {
+      this.inputMin.value = String(minValue);
+    }
+    if (this.inputMax) {
+      this.inputMax.value = String(maxValue);
+    }
   }
 
-  updated(): void {
-    this.syncNativeInputValues();
+  protected ensureValues(
+    properties: PropertyValues<MultiRangeProperties>
+  ): void {
+    if (this.hasInvalidRange() || !this.hasDiffs(properties)) return;
 
-    this.dispatchEvent(
-      new CustomEvent<MultiRangeChangeEvent>('change', {
-        bubbles: true,
-        composed: true,
-        detail: {
-          minValue: this.minValue,
-          maxValue: this.maxValue,
-        },
-      })
-    );
+    let minValue = this.minValue!;
+    let maxValue = this.maxValue!;
+    const min = this.min!;
+    const max = this.max!;
+
+    const minInRange = minValue >= min && minValue <= max - this.step;
+    const maxInRange = maxValue >= min + this.step && maxValue <= max;
+    const minBeforeMax = minValue <= maxValue - this.step;
+
+    if (!minInRange || !minBeforeMax) minValue = min;
+    if (!maxInRange || !minBeforeMax) maxValue = max;
+
+    this.minValue = minValue;
+    this.maxValue = maxValue;
+
+    this.setPercentages(minValue, maxValue);
+    this.syncNativeInputValues(minValue, maxValue);
   }
 
-  protected syncNativeInputValues(): void {
-    this.minValue = Number(this.inputMinRef?.value?.value ?? this.minValue);
-    this.maxValue = Number(this.inputMaxRef?.value?.value ?? this.maxValue);
+  protected hasDiffs(
+    properties: PropertyValues<MultiRangeProperties>
+  ): boolean {
+    return Array.from(properties.entries()).some(([_key, value]) => {
+      const key = _key as keyof MultiRangeComponent;
+      return this.observedKeys.includes(key) && this[key] !== value;
+    });
   }
 
-  protected override render(): TemplateResult {
+  protected hasInvalidRange(): boolean {
+    this.invalid =
+      this.observedKeys.some((key) => typeof this[key] === 'undefined') ||
+      this.min! > this.max! - this.step ||
+      this.max! < this.min! + this.step;
+
+    return this.invalid;
+  }
+
+  protected override render(): TemplateResult | void {
+    if (this.invalid) return;
+
     return html`
-      ${this.renderRangeInput(this.minValue, this.inputMinRef, true)}
-      ${this.renderRangeInput(this.maxValue, this.inputMaxRef)}
-      <div class="active"></div>
+      ${this.renderRangeInput(this._min!, true)}
+      ${this.renderRangeInput(this._max!)}
     `;
   }
 
-  protected renderRangeInput(
-    value: number,
-    inputRef: Ref,
-    isFirst = false
-  ): TemplateResult {
+  protected renderRangeInput(value: number, isFirst = false): TemplateResult {
```

Comment:
It's a breaking change, as long as we changed method signature.

Reply:
refactored